### PR TITLE
NIP-118, Polls on Nostr 

### PR DESCRIPTION
**Pull request by abhay-raizada**
This NIP takes the ideas evolved from 
- https://github.com/nostr-protocol/nips/pull/1190
- https://github.com/nostr-protocol/nips/pull/1324
- https://github.com/vitorpamplona/amethyst/issues/935
- https://github.com/nostr-protocol/nips/pull/320
- https://github.com/nostr-protocol/nips/pull/148

and tries to combine them in a spec for a  simple polls implementation. 

Working version: https://pollerama.fun/

cc: @staab @vitorpamplona @vicariousdrama @alexgleason @gocedoko  @fiatjaf

### CODE REVIEW

File: 118.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+The NIP makes the following tradeoffs.
+
+1. All votes are public.
+2. All responses are editable (latest by timestamp considered last response)
+
+## Events
+
+### Poll Content
+
+The poll event is defined as a `kind:1068` event. Major tags in the poll event are:
```
Comment (AsaiToshiya) [inline 118.md:18]: I think defining the new kind is a good point.

File: 118.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+- **label**: The accompanying text of the poll.
```
Comment (AsaiToshiya) [inline 118.md:20]: Why not use the `content`?

File: 118.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+- **option**: The option tags contain an OptionId(any alphanumeric) field, followed by an option label field.
```
Comment (AsaiToshiya) [inline 118.md:21]: I don't understand why you need the OptionId. The label can be the ID.

File: 118.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+- **relays**: Set of relays that the poll is expecting respondents to respond on.
+- **settings**: A JSON stringified object that may hold poll level configs and client specific configs.
```
Comment (AsaiToshiya) [inline 118.md:23]: Why the JSON? I haven't seen it as the value of the tag.

File: 118.md (hunk 3 above)
Comment (abhay-raizada) [reply 118.md:21]: what if there are two labels with the same name? i think this is an easy way to get past that. I don't like the idea of tallying same labels together.

File: 118.md (hunk 2 above)
Comment (abhay-raizada) [reply 118.md:20]: it makes it easier if we want to take the same polls format and encrypt it as a private poll, There are no private polls yet however, so i'm open to move this to a label

File: 118.md (hunk 4 above)
Comment (abhay-raizada) [reply 118.md:23]: carried it over from forms where i use it for Ui settings like imageurl etc. Thought I might reuse the same config for the pollType field, but I think the PollType field in the tags also makes sense.

File: 118.md (hunk 3 above)
Comment (AsaiToshiya) [reply 118.md:21]: I feel like a label with the same name but different meaning is not a practical option.

Comment (vitorpamplona) [reply 118.md:21]: I agree with @AsaiToshiya. If clients are only showing the name and not the option ID, the same name should mean the same vote. Not having an ID removes the need to disambiguate names when they duplicate.

File: 118.md (hunk 4 above)
Comment (vitorpamplona) [reply 118.md:23]: I would get rid of it. Add all the extra settings as tags.

File: 118.md (hunk 2 above)
Comment (AsaiToshiya) [reply 118.md:20]: I prefer the `content` unless the specification for private polls is clear.

File: 118.md (hunk 3 above)
Comment (abhay-raizada) [reply 118.md:21]: I disagree, the client is not doing anything in this special case.

Its the poll creator that is just having fun, maybe they want to see how the option order affects the poll results, which they should be allowed to do.

Comment (mikedilger) [reply 118.md:21]: I agree with abhay-raizada.  I might make a poll that says:  Who will you vote for?  1) Trump, 2) Trump, 3) Trump
and it should collect the three answers separately.

File: 118.md [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  "created_at": 1719888496,
+  "id": "9d1b6b9562e66f2ecf35eb0a3c2decc736c47fddb13d6fb8f87185a153ea3634",
+  "kind": 1068,
+  "pubkey": "dee45a23c4f1d93f3a2043650c5081e4ac14a778e0acbef03de3768e4f81ac7b",
+  "sig": "7fa93bf3c430eaef784b0dacc217d3cd5eff1c520e7ef5d961381bc0f014dde6286618048d924808e54d1be03f2f2c2f0f8b5c9c2082a4480caf45a565ca9797",
+  "tags": [
+    ["label", "Pineapple on Pizza"],
+    ["option", "qj518h583", "Yay"],
+    ["option", "gga6cdnqj", "Nay"],
+    ["relays", "<relay1 url>", "<relay 2 url>"],
```
[the remark this answers is not in the document]
Comment (staab) [reply 118.md:39]: Yeah, variadic tags are usually bad

File: 118.md [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+}
+```
+
+### Poll Types
+
+Poll Types can be configured in the settings object of the poll, the setting dictates how multiple response tags are handled in the `kind:1018` event.
+
+- **PollType: singlechoice**: The first response tag is to be considered the actual response.
+- **PollType: multiplechoice**: The first response tag pointing to each id is considered the actual response, without considering the order.
+- **PollType: rankedchoice**: The first response tag pointing to each id is the considered the actual response, while also taking into account the order of the response tags.
```
Comment (staab) [inline 118.md:74]: Instead of JSONified `settings`, settings should just be tags

File: 118.md [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+In case of multiple events for a pubkey, the event with the largest timestamp should be considered.
+
+### Curation
+
+The clients may configure fetching results by specific people. This can be achieved by creating `kind:30000` follow sets, and fetching results only from the follow set.
+Clients can also employ other curation algorithms, like Proof Of Work and Web of Trust scores for result curations.
+
+### Private Polls
+
+This nip does not cover private polling, Clients may follow [forms nip](https://github.com/nostr-protocol/nips/pull/1190) for private polling.
```
Comment (staab) [inline 118.md:89]: I'd omit this section, it's not necessary and nips shouldn't refer to PRs

**Review by staab — APPROVED** [review]
This is the best poll nip so far. A few minor changes, but it's the right level of dumb simplicity.

**Comment by alexgleason** [comment]
Does this have 2 implementations? I was just about to comment on it:

- Why is response not a parameterized replaceable event?
- Why not use the tag index for the vote instead of making up random letters?

Not using a replaceable event for the response is especially bad if the response is meant to be edited.

**Comment by alexgleason** [comment]
I will actually implement polls. I have a UI for it and could do it in a couple hours. But counting votes is going to be such a pain because we don't use existing replaceable semantics.

**Comment by fiatjaf** [comment]
> * Why is response not a parameterized replaceable event?
> * Why not use the tag index for the vote instead of making up random letters?

I asked exactly these two questions myself, but I decided it wasn't a big deal.

File: 118.md [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+```json
+{
+  "content": "",
+  "created_at": 1719888496,
+  "id": "9d1b6b9562e66f2ecf35eb0a3c2decc736c47fddb13d6fb8f87185a153ea3634",
+  "kind": 1068,
+  "pubkey": "dee45a23c4f1d93f3a2043650c5081e4ac14a778e0acbef03de3768e4f81ac7b",
+  "sig": "7fa93bf3c430eaef784b0dacc217d3cd5eff1c520e7ef5d961381bc0f014dde6286618048d924808e54d1be03f2f2c2f0f8b5c9c2082a4480caf45a565ca9797",
+  "tags": [
+    ["label", "Pineapple on Pizza"],
```
Comment (smolgrrr) [inline 118.md:41]: implemented NIP-88 on getwired.app
super easy to implement and works well. don't really see the point of the label tag on the poll event, but other than that no complaints

Comment (abhay-raizada) [reply 118.md:41]: it's an artifact from before the label was moved to the content field. Will update

**Pull request merged**